contracts/Uniswap/UniswapAnchoredView.sol

```diff
@@ -228,103 +198,71 @@ contract UniswapAnchoredView is AggregatorValidatorInterface, UniswapConfig, Own
     }
 
     /**
-     * @dev Fetches the current token/eth price accumulator from uniswap.
+     * @dev Fetches the latest TWAP from the UniV3 pool tick, over the last anchor period.
      */
-    function currentCumulativePrice(TokenConfig memory config) internal view returns (uint) {
-        (uint cumulativePrice0, uint cumulativePrice1,) = UniswapV2OracleLibrary.currentCumulativePrices(config.uniswapMarket);
+    function getUniswapTwap(TokenConfig memory config) internal view returns (uint256) {
+        uint32 anchorPeriod_ = uint32(anchorPeriod); // overflow is expected
+        uint32[] memory secondsAgos = new uint32[](2);
+        secondsAgos[0] = anchorPeriod_;
+        secondsAgos[1] = 0;
+        (int56[] memory tickCumulatives, ) = IUniswapV3Pool(config.uniswapMarket).observe(secondsAgos);
+
+        int24 timeWeightedAverageTick = int24((tickCumulatives[1] - tickCumulatives[0]) / anchorPeriod_);
```

Comment:
This seems dangerous. Is there a reason we're using raw arithmetic rather than some form of SafeMath?

Can the tick be negative? How does the tick being positive or negative affect the price?

Reply:
Yes - these are the minimum and maximum tick values for a Uni V3 Pool: https://github.com/Uniswap/uniswap-v3-core/blob/main/contracts/libraries/TickMath.sol#L9
Price is calculated as a function of the tick as the exponent, like `price = 1.0001^{tick}`. So increasingly negative ticks means price (relative to the other token in the pool) tends towards zero.